packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,211 @@
+      throw new Error(`Keyring: Could not generate address in group ${group}, group is invalid`)
+
+    if (addressIndex !== undefined) {
+      if (!Number.isInteger(addressIndex) || addressIndex < 0)
+        throw new Error(`Keyring: Could not generate address, ${addressIndex} is not a valid addressIndex`)
+
+      if (group !== undefined || skipAddressIndexes.length > 0)
+        throw new Error(
+          'Keyring: Could not generate address, invalid arguments passed: when addressIndex is provided the group and skipAddressIndexes should not be provided.'
+        )
+
+      return this._deriveAddressAndKeys(addressIndex)
+    }
+
+    const initialAddressIndex = 0
+
+    let nextAddressIndex = skipAddressIndexes.includes(initialAddressIndex)
+      ? findNextAvailableAddressIndex(initialAddressIndex, skipAddressIndexes)
+      : initialAddressIndex
+    let newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+
+    while (group !== undefined && groupOfAddress(newAddressData.hash) !== group) {
+      nextAddressIndex = findNextAvailableAddressIndex(newAddressData.index, skipAddressIndexes)
+      newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+    }
+
+    return newAddressData
+  }
+
+  private _getNonSensitiveAddressData = ({
+    hash,
+    index,
+    publicKey
+  }: SensitiveAddressData): NonSensitiveAddressData => ({ hash, index, publicKey })
+
+  private _initFromMnemonic = (mnemonic: Uint8Array | null, passphrase: string) => {
+    if (this.root) throw new Error('Keyring: Secret recovery phrase already provided')
+    if (!mnemonic) throw new Error('Keyring: Secret recovery phrase not provided')
+
+    const seed = metamaskBip39.mnemonicToSeedSync(mnemonic, wordlist, passphrase)
+    this.hdWallet = HDKey.fromMasterSeed(seed)
+    this.root = this.hdWallet.derive(this.hdPath)
+
+    passphrase = ''
+  }
+
+  private _deriveAddressAndKeys = (addressIndex: number): SensitiveAddressData => {
+    if (!this.root) throw new Error('Keyring: Cannot derive address and keys, secret recovery phrase is not provided')
+    if (!isAddressIndexValid(addressIndex)) throw new Error('Invalid address index path level')
+
+    const keyPair = this.root.deriveChild(addressIndex)
+
+    if (!keyPair.publicKey) throw new Error('Keyring: Missing public key')
+    if (!keyPair.privateKey) throw new Error('Keyring: Missing private key')
+
+    const publicKey = bytesToHex(keyPair.publicKey)
+    const privateKey = keyPair.privateKey
+    const hash = blake.blake2b(Uint8Array.from(keyPair.publicKey), undefined, 32)
+    const type = new Uint8Array([0])
+    const bytes = new Uint8Array(type.length + hash.length)
+    bytes.set(type, 0)
+    bytes.set(hash, type.length)
+    const address = bs58.encode(bytes)
+
+    return { hash: address, publicKey, privateKey, index: addressIndex }
+  }
+}
+
+export const keyring = new Keyring()
```

Comment:
It is in this class instance that the secrets (`root`, `privateKey`) are cached, as per Metamask: https://github.com/MetaMask/eth-hd-keyring/blob/main/index.js